rust/src/io/object_reader.rs

```diff
@@ -63,15 +63,23 @@ pub struct CloudObjectReader {
     pub path: Path,
 
     block_size: usize,
+
+    cache: Arc<dyn ObjectCache>,
 }
 
 impl<'a> CloudObjectReader {
     /// Create an ObjectReader from URI
-    pub fn new(object_store: &'a ObjectStore, path: Path, block_size: usize) -> Result<Self> {
+    pub fn new(
+        object_store: &'a ObjectStore,
+        path: Path,
+        block_size: usize,
+        cache: Arc<dyn ObjectCache>,
```

Comment:
The main question is why `ObjectCache` is a composite member of ObjectReader, instead of 

```
impl ObjectStore for CachedObjectStore {}
```

Also, not clear to me why each object cache is bounded to each file?

Reply:
this is to make sure we can read `get_range` api so we can cache at finer granularity in the future. Let make make some changes. so caching is less intrusive